xapian-core/api/registry.cc

```diff
@@ -260,13 +322,16 @@ void
 Registry::register_weighting_scheme(const Xapian::Weight &wt)
 {
     LOGCALL_VOID(API, "Xapian::Registry::register_weighting_scheme", wt.name());
-    register_object(internal->wtschemes, wt);
+    register_object(internal->wtschemes, internal->wtschemes_short, wt);
 }
 
 const Xapian::Weight *
 Registry::get_weighting_scheme(const string & name) const
 {
     LOGCALL(API, const Xapian::Weight *, "Xapian::Registry::get_weighting_scheme", name);
+    const char * p = name.c_str();
+    if (std::islower(*p))
+	RETURN(lookup_object(internal->wtschemes_short, name));
```

Comment:
Use `C_islower()` rather than `std::islower()` (and drop the `<cctype>` include).  Functions like `std::islower()` change behaviour depending on the currently set locale (because the definition of what a lower case letter is varies by language), and user code may well change the current locale.  To avoid behaviour depending on the current locale, xapian-core has its own `C_islower()`, etc which act like `std::islower()` in the "C" locale (i.e. as if `setlocale(LC_CTYPE, "C");` were in effect).

Reusing `get_weighting_scheme()` for both lookups like this seems OK, but means that the long name really mustn't begin with a lower case letter.  While that is the current convention, I don't think it's documented as a requirement and certainly isn't currently enforced in any way, so perhaps if the short name lookup returns `NULL` we should continue on and try the long name lookup, to avoid needlessly breaking any current user subclasses that return a lower case name.

`name[0]` is simpler and clearer than using `c_str()` and then `*`, and avoids an extra variable.  But the code should also be robust to being passed an empty string.  So something like `!name.empty() && C_islower(name[0])`